neuralmonkey/attention/scaled_dot_product.py

```diff
@@ -23,6 +24,143 @@
 # pylint: enable=invalid-name
 
 
+def split_for_heads(x: tf.Tensor, n_heads: int, head_dim: int) -> tf.Tensor:
+    """Split a tensor for multi-head attention.
+
+    Split last dimension of 3D vector of shape (batch, time, dim) and return
+    a 4D vector with shape (batch, n_heads, time, dim/n_heads)
+    """
+    x_shape = tf.shape(x)
+    x_4d = tf.reshape(tf.expand_dims(x, 2),
+                      [x_shape[0], x_shape[1], n_heads, head_dim])
+
+    return tf.transpose(x_4d, perm=[0, 2, 1, 3])
+
+
+def mask_weights(weights_4d: tf.Tensor, mask: tf.Tensor) -> tf.Tensor:
+    """Apply mask to softmax weights and renormalize."""
+    # attention mask shape: batch, time(k)
+    # weights_all shape: batch, head, time(q), time(k)
+    weights_all = weights_4d * tf.expand_dims(tf.expand_dims(mask, 1), 1)
+
+    # normalization along time(k)
+    # norm shape: batch, head, time(q), 1
+    norm = tf.reduce_sum(weights_all, 3, keep_dims=True) + 1e-8
+    return weights_all / norm
+
+
+def mask_future(energies: tf.Tensor) -> tf.Tensor:
+    """Mask energies of keys to the right of the query."""
+    triangular_mask = tf.matrix_band_part(tf.ones_like(energies), -1, 0)
+    mask_area = tf.equal(triangular_mask, 1)
+    masked_value = tf.fill(tf.shape(energies), -np.inf)
+    return tf.where(mask_area, energies, masked_value)
+
+
+def attention(
+        queries: tf.Tensor,
+        keys: tf.Tensor,
+        values: tf.Tensor,
+        keys_mask: tf.Tensor,
+        num_heads: int,
+        dropout_callback: Callable[[tf.Tensor], tf.Tensor],
+        masked: bool = False) -> tf.Tensor:
```

Comment:
Docstring:
Run multi-head scaled dot-product attention.

do ``Args:`` a ``Returns:`` přesunout ty shapy co jsou v komentářích o pár řádků níž, plus:
``keys_mask`` a float Tensor for masking sequences in keys
``num_heads`` how many attention heads to use
``dropout_callback`` a function that performs dropout (so that the `train_mode` and `dropout_keep_prob` can be "closured" in)
``masked`` whether to apply triangular mask to the queries (handy in autoregressive decoders, used to mask-out the future states)